// @nrwl/webpack/src/utils/webpack/partials/browser.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getBrowserConfig = void 0;
const license_webpack_plugin_1 = require("license-webpack-plugin");
const webpack_1 = require("webpack");
const webpack_subresource_integrity_1 = require("webpack-subresource-integrity");
const CssMinimizerPlugin = require("css-minimizer-webpack-plugin");
function getBrowserConfig(wco) {
    const { buildOptions, supportES2015 } = wco;
    const extraPlugins = [];
    const stylesOptimization = typeof buildOptions.optimization === 'object'
        ? buildOptions.optimization.styles
        : buildOptions.optimization;
    if (buildOptions.subresourceIntegrity) {
        extraPlugins.push(new webpack_subresource_integrity_1.SubresourceIntegrityPlugin());
    }
    if (buildOptions.extractLicenses) {
        extraPlugins.push(new license_webpack_plugin_1.LicenseWebpackPlugin({
            stats: {
                warnings: false,
                errors: false,
            },
            perChunkOutput: false,
            outputFilename: `3rdpartylicenses.txt`,
        }));
    }
    const extraMinimizers = [];
    if (stylesOptimization) {
        extraMinimizers.push(new CssMinimizerPlugin({
            test: /\.(?:css|scss|sass|less|styl)$/,
        }));
    }
    return {
        resolve: {
            mainFields: [
                ...(supportES2015 ? ['es2015'] : []),
                'browser',
                'module',
                'main',
            ],
        },
        output: {
            crossOriginLoading: buildOptions.subresourceIntegrity
                ? 'anonymous'
                : false,
        },
        // context needs to be set for babel to pick up correct babelrc
        context: wco.projectRoot,
        optimization: {
            minimizer: [new webpack_1.ids.HashedModuleIdsPlugin(), ...extraMinimizers],
            emitOnErrors: false,
            moduleIds: 'deterministic',
            runtimeChunk: buildOptions.runtimeChunk ? 'single' : false,
            splitChunks: {
                maxAsyncRequests: Infinity,
                cacheGroups: {
                    default: !!buildOptions.commonChunk && {
                        chunks: 'async',
                        minChunks: 2,
                        priority: 10,
                    },
                    common: !!buildOptions.commonChunk && {
                        name: 'common',
                        chunks: 'async',
                        minChunks: 2,
                        enforce: true,
                        priority: 5,
                    },
                    vendors: false,
                    vendor: !!buildOptions.vendorChunk && {
                        name: 'vendor',
                        chunks: (chunk) => chunk.name === 'main',
                        enforce: true,
                        test: /[\\/]node_modules[\\/]/,
                    },
                },
            },
        },
        plugins: extraPlugins,
        node: false,
    };
}
exports.getBrowserConfig = getBrowserConfig;
