// @nrwl/storybook/src/utils/versions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.litHtmlVersion = exports.storybookTestRunnerVersion = exports.storybookNextAddonVersion = exports.storybookSwcAddonVersion = exports.reactNativeStorybookLoader = exports.storybookReactNativeVersion = exports.webpack5Version = exports.urlLoaderVersion = exports.svgrVersion = exports.babelPresetTypescriptVersion = exports.babelLoaderVersion = exports.babelCoreVersion = exports.storybookVersion = exports.nxVersion = void 0;
exports.nxVersion = require('../../package.json').version;
exports.storybookVersion = '~6.5.9';
exports.babelCoreVersion = '7.12.13';
exports.babelLoaderVersion = '8.1.0';
exports.babelPresetTypescriptVersion = '7.12.13';
exports.svgrVersion = '^5.4.0';
exports.urlLoaderVersion = '^3.0.0';
exports.webpack5Version = '^5.64.0';
exports.storybookReactNativeVersion = '^6.0.1-beta.5';
exports.reactNativeStorybookLoader = '^2.0.5';
exports.storybookSwcAddonVersion = '^1.1.7';
exports.storybookNextAddonVersion = '^1.6.6';
exports.storybookTestRunnerVersion = '^0.7.2';
exports.litHtmlVersion = '^2.3.1';
