// @nrwl/next/src/generators/application/lib/add-project.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.addProject = void 0;
const devkit_1 = require("@nrwl/devkit");
function addProject(host, options) {
    const targets = {};
    targets.build = {
        builder: '@nrwl/next:build',
        outputs: ['{options.outputPath}'],
        defaultConfiguration: 'production',
        options: {
            root: options.appProjectRoot,
            outputPath: (0, devkit_1.joinPathFragments)('dist', options.appProjectRoot),
        },
        configurations: {
            development: {
                outputPath: options.appProjectRoot,
            },
            production: {},
        },
    };
    targets.serve = {
        builder: '@nrwl/next:server',
        defaultConfiguration: 'development',
        options: {
            buildTarget: `${options.projectName}:build`,
            dev: true,
        },
        configurations: {
            development: {
                buildTarget: `${options.projectName}:build:development`,
                dev: true,
            },
            production: {
                buildTarget: `${options.projectName}:build:production`,
                dev: false,
            },
        },
    };
    if (options.server) {
        targets.serve.options = Object.assign(Object.assign({}, targets.serve.options), { customServerPath: options.server });
    }
    targets.export = {
        builder: '@nrwl/next:export',
        options: {
            buildTarget: `${options.projectName}:build:production`,
        },
    };
    const project = {
        root: options.appProjectRoot,
        sourceRoot: options.appProjectRoot,
        projectType: 'application',
        targets,
        tags: options.parsedTags,
    };
    (0, devkit_1.addProjectConfiguration)(host, options.projectName, Object.assign({}, project), options.standaloneConfig);
}
exports.addProject = addProject;
