// @nrwl/react/src/migrations/update-14-0-0/update-react-dom-render-for-v18.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.migrateReactDomRender = exports.update = void 0;
const tslib_1 = require("tslib");
const devkit_1 = require("@nrwl/devkit");
const find_nodes_1 = require("@nrwl/workspace/src/utilities/typescript/find-nodes");
const ts = require("typescript");
function update(tree) {
    return tslib_1.__awaiter(this, void 0, void 0, function* () {
        const projects = (0, devkit_1.getProjects)(tree);
        projects.forEach((config, name) => {
            var _a, _b;
            const isReactProject = ((_b = (_a = config.targets) === null || _a === void 0 ? void 0 : _a.build) === null || _b === void 0 ? void 0 : _b.executor) === '@nrwl/web:webpack' &&
                /main\.(t|j)sx?$/.test(config.targets.build.options.main);
            if (isReactProject) {
                const sourcePath = config.targets.build.options.main;
                const sourceCode = tree.read(sourcePath).toString();
                const source = ts.createSourceFile(sourcePath, sourceCode, ts.ScriptTarget.Latest, true);
                const result = (0, devkit_1.applyChangesToString)(sourceCode, migrateReactDomRender(sourcePath, source));
                tree.write(sourcePath, result);
            }
        });
        yield (0, devkit_1.formatFiles)(tree);
    });
}
exports.update = update;
function migrateReactDomRender(sourcePath, source) {
    const allImports = (0, find_nodes_1.findNodes)(source, ts.SyntaxKind.ImportDeclaration);
    const reactDomImport = allImports.find((x) => x.moduleSpecifier.getText() === "'react-dom'");
    const changes = [];
    if (reactDomImport) {
        changes.push({
            type: devkit_1.ChangeType.Insert,
            index: reactDomImport.moduleSpecifier.end - 1,
            text: '/client',
        });
    }
    const calls = (0, find_nodes_1.findNodes)(source, ts.SyntaxKind.CallExpression);
    const renderCall = calls.find((x) => {
        if (x.expression.kind !== ts.SyntaxKind.PropertyAccessExpression)
            return false;
        const expr = x.expression;
        return (expr.expression.getText() === 'ReactDOM' &&
            expr.name.getText() === 'render');
    });
    if (renderCall) {
        const [element, querySelector] = renderCall.arguments;
        changes.push({
            type: devkit_1.ChangeType.Delete,
            start: renderCall.getStart(),
            length: renderCall.end,
        }, {
            type: devkit_1.ChangeType.Insert,
            index: renderCall.getStart(),
            text: (0, devkit_1.stripIndents) `
          const root = ReactDOM.createRoot(
            ${querySelector.getText()}${sourcePath.endsWith('.tsx') ? ' as HTMLElement' : ''}
          );
          root.render(
            ${element.getText()}
          );
        `,
        });
    }
    return changes;
}
exports.migrateReactDomRender = migrateReactDomRender;
exports.default = update;
