// @nrwl/react/src/generators/component/component.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.componentSchematic = exports.componentGenerator = void 0;
const tslib_1 = require("tslib");
const devkit_1 = require("@nrwl/devkit");
const run_tasks_in_serial_1 = require("@nrwl/workspace/src/utilities/run-tasks-in-serial");
const ts = require("typescript");
const add_styled_dependencies_1 = require("../../rules/add-styled-dependencies");
const assertion_1 = require("../../utils/assertion");
const ast_utils_1 = require("../../utils/ast-utils");
const versions_1 = require("../../utils/versions");
function componentGenerator(host, schema) {
    return tslib_1.__awaiter(this, void 0, void 0, function* () {
        const options = yield normalizeOptions(host, schema);
        createComponentFiles(host, options);
        const tasks = [];
        const styledTask = (0, add_styled_dependencies_1.addStyledModuleDependencies)(host, options.styledModule);
        tasks.push(styledTask);
        addExportsToBarrel(host, options);
        if (options.routing) {
            const routingTask = (0, devkit_1.addDependenciesToPackageJson)(host, { 'react-router-dom': versions_1.reactRouterDomVersion }, { '@types/react-router-dom': versions_1.typesReactRouterDomVersion });
            tasks.push(routingTask);
        }
        yield (0, devkit_1.formatFiles)(host);
        return (0, run_tasks_in_serial_1.runTasksInSerial)(...tasks);
    });
}
exports.componentGenerator = componentGenerator;
function createComponentFiles(host, options) {
    const componentDir = (0, devkit_1.joinPathFragments)(options.projectSourceRoot, options.directory);
    (0, devkit_1.generateFiles)(host, (0, devkit_1.joinPathFragments)(__dirname, './files'), componentDir, Object.assign(Object.assign({}, options), { tmpl: '' }));
    for (const c of host.listChanges()) {
        let deleteFile = false;
        if (options.skipTests && /.*spec.tsx/.test(c.path)) {
            deleteFile = true;
        }
        if ((options.styledModule || !options.hasStyles) &&
            c.path.endsWith(`.${options.style}`)) {
            deleteFile = true;
        }
        if (options.globalCss && c.path.endsWith(`.module.${options.style}`)) {
            deleteFile = true;
        }
        if (!options.globalCss &&
            c.path.endsWith(`${options.fileName}.${options.style}`)) {
            deleteFile = true;
        }
        if (deleteFile) {
            host.delete(c.path);
        }
    }
    if (options.js) {
        (0, devkit_1.toJS)(host);
    }
}
function addExportsToBarrel(host, options) {
    const workspace = (0, devkit_1.getProjects)(host);
    const isApp = workspace.get(options.project).projectType === 'application';
    if (options.export && !isApp) {
        const indexFilePath = (0, devkit_1.joinPathFragments)(options.projectSourceRoot, options.js ? 'index.js' : 'index.ts');
        const indexSource = host.read(indexFilePath, 'utf-8');
        if (indexSource !== null) {
            const indexSourceFile = ts.createSourceFile(indexFilePath, indexSource, ts.ScriptTarget.Latest, true);
            const changes = (0, devkit_1.applyChangesToString)(indexSource, (0, ast_utils_1.addImport)(indexSourceFile, `export * from './${options.directory}/${options.fileName}';`));
            host.write(indexFilePath, changes);
        }
    }
}
function normalizeOptions(host, options) {
    var _a, _b, _c, _d;
    return tslib_1.__awaiter(this, void 0, void 0, function* () {
        assertValidOptions(options);
        const { className, fileName } = (0, devkit_1.names)(options.name);
        const componentFileName = (_a = options.fileName) !== null && _a !== void 0 ? _a : (options.pascalCaseFiles ? className : fileName);
        const project = (0, devkit_1.getProjects)(host).get(options.project);
        if (!project) {
            devkit_1.logger.error(`Cannot find the ${options.project} project. Please double check the project name.`);
            throw new Error();
        }
        const { sourceRoot: projectSourceRoot, projectType } = project;
        const directory = yield getDirectory(host, options);
        const styledModule = /^(css|scss|less|styl|none)$/.test(options.style)
            ? null
            : options.style;
        if (options.export && projectType === 'application') {
            devkit_1.logger.warn(`The "--export" option should not be used with applications and will do nothing.`);
        }
        options.classComponent = (_b = options.classComponent) !== null && _b !== void 0 ? _b : false;
        options.routing = (_c = options.routing) !== null && _c !== void 0 ? _c : false;
        options.globalCss = (_d = options.globalCss) !== null && _d !== void 0 ? _d : false;
        return Object.assign(Object.assign({}, options), { directory,
            styledModule, hasStyles: options.style !== 'none', className, fileName: componentFileName, projectSourceRoot });
    });
}
function getDirectory(host, options) {
    return tslib_1.__awaiter(this, void 0, void 0, function* () {
        const genNames = (0, devkit_1.names)(options.name);
        const fileName = options.pascalCaseDirectory === true
            ? genNames.className
            : genNames.fileName;
        const workspace = (0, devkit_1.getProjects)(host);
        let baseDir;
        if (options.directory) {
            baseDir = options.directory;
        }
        else {
            baseDir =
                workspace.get(options.project).projectType === 'application'
                    ? 'app'
                    : 'lib';
        }
        return options.flat ? baseDir : (0, devkit_1.joinPathFragments)(baseDir, fileName);
    });
}
function assertValidOptions(options) {
    (0, assertion_1.assertValidStyle)(options.style);
    const slashes = ['/', '\\'];
    slashes.forEach((s) => {
        if (options.name.indexOf(s) !== -1) {
            const [name, ...rest] = options.name.split(s).reverse();
            let suggestion = rest.map((x) => x.toLowerCase()).join(s);
            if (options.directory) {
                suggestion = `${options.directory}${s}${suggestion}`;
            }
            throw new Error(`Found "${s}" in the component name. Did you mean to use the --directory option (e.g. \`nx g c ${name} --directory ${suggestion}\`)?`);
        }
    });
}
exports.default = componentGenerator;
exports.componentSchematic = (0, devkit_1.convertNxGenerator)(componentGenerator);
