// @nrwl/jest/src/generators/init/init.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.jestInitSchematic = exports.jestInitGenerator = void 0;
const devkit_1 = require("@nrwl/devkit");
const versions_1 = require("../../utils/versions");
const schemaDefaults = {
    compiler: 'tsc',
    js: false,
};
function createJestConfig(tree, js = false) {
    // if the root ts config already exists then don't make a js one or vice versa
    if (!tree.exists('jest.config.ts') && !tree.exists('jest.config.js')) {
        const contents = js
            ? (0, devkit_1.stripIndents) `
      const { getJestProjects } = require('@nrwl/jest');

      module.exports = {
        projects: getJestProjects()
      };`
            : (0, devkit_1.stripIndents) `
      import { getJestProjects } from '@nrwl/jest';

      export default {
       projects: getJestProjects()
      };`;
        tree.write(`jest.config.${js ? 'js' : 'ts'}`, contents);
    }
    if (!tree.exists('jest.preset.js')) {
        // preset is always js file.
        tree.write(`jest.preset.js`, `
      const nxPreset = require('@nrwl/jest/preset').default;
     
      module.exports = { ...nxPreset }`);
        addTestInputs(tree);
    }
}
function addTestInputs(tree) {
    var _a, _b, _c, _d;
    var _e, _f;
    const workspaceConfiguration = (0, devkit_1.readWorkspaceConfiguration)(tree);
    const productionFileSet = (_a = workspaceConfiguration.namedInputs) === null || _a === void 0 ? void 0 : _a.production;
    if (productionFileSet) {
        // This is one of the patterns in the default jest patterns
        productionFileSet.push(
        // Remove spec, test, and snapshots from the production fileset
        '!{projectRoot}/**/?(*.)+(spec|test).[jt]s?(x)?(.snap)', 
        // Remove tsconfig.spec.json
        '!{projectRoot}/tsconfig.spec.json', 
        // Remove jest.config.js/ts
        '!{projectRoot}/jest.config.[jt]s');
        // Dedupe and set
        workspaceConfiguration.namedInputs.production = Array.from(new Set(productionFileSet));
    }
    // Test targets depend on all their project's sources + production sources of dependencies
    (_b = workspaceConfiguration.targetDefaults) !== null && _b !== void 0 ? _b : (workspaceConfiguration.targetDefaults = {});
    (_c = (_e = workspaceConfiguration.targetDefaults).test) !== null && _c !== void 0 ? _c : (_e.test = {});
    (_d = (_f = workspaceConfiguration.targetDefaults.test).inputs) !== null && _d !== void 0 ? _d : (_f.inputs = [
        'default',
        productionFileSet ? '^production' : '^default',
    ]);
    workspaceConfiguration.targetDefaults.test.inputs.push('{workspaceRoot}/jest.preset.js');
    (0, devkit_1.updateWorkspaceConfiguration)(tree, workspaceConfiguration);
}
function updateDependencies(tree, options) {
    const dependencies = {
        tslib: versions_1.tslibVersion,
    };
    const devDeps = {
        '@nrwl/jest': versions_1.nxVersion,
        jest: versions_1.jestVersion,
        'jest-environment-jsdom': versions_1.jestVersion,
        // because the default jest-preset uses ts-jest,
        // jest will throw an error if it's not installed
        // even if not using it in overriding transformers
        'ts-jest': versions_1.tsJestVersion,
    };
    if (!options.js) {
        devDeps['ts-node'] = versions_1.tsNodeVersion;
        devDeps['@types/jest'] = versions_1.jestTypesVersion;
        devDeps['@types/node'] = '16.11.7';
    }
    if (options.compiler === 'babel' || options.babelJest) {
        devDeps['babel-jest'] = versions_1.babelJestVersion;
        // in some cases @nrwl/web will not already be present i.e. node only projects
        devDeps['@nrwl/web'] = versions_1.nxVersion;
    }
    else if (options.compiler === 'swc') {
        devDeps['@swc/jest'] = versions_1.swcJestVersion;
    }
    return (0, devkit_1.addDependenciesToPackageJson)(tree, dependencies, devDeps);
}
function updateExtensions(host) {
    if (!host.exists('.vscode/extensions.json')) {
        return;
    }
    (0, devkit_1.updateJson)(host, '.vscode/extensions.json', (json) => {
        json.recommendations = json.recommendations || [];
        const extension = 'firsttris.vscode-jest-runner';
        if (!json.recommendations.includes(extension)) {
            json.recommendations.push(extension);
        }
        return json;
    });
}
function jestInitGenerator(tree, schema) {
    const options = normalizeOptions(schema);
    createJestConfig(tree, options.js);
    let installTask = () => { };
    if (!options.skipPackageJson) {
        (0, devkit_1.removeDependenciesFromPackageJson)(tree, ['@nrwl/jest'], []);
        installTask = updateDependencies(tree, options);
    }
    updateExtensions(tree);
    return installTask;
}
exports.jestInitGenerator = jestInitGenerator;
function normalizeOptions(options) {
    return Object.assign(Object.assign({}, schemaDefaults), options);
}
exports.default = jestInitGenerator;
exports.jestInitSchematic = (0, devkit_1.convertNxGenerator)(jestInitGenerator);
