// @nrwl/webpack/src/executors/webpack/lib/get-emitted-files.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getEmittedFiles = void 0;
const path_1 = require("path");
function getEmittedFiles(stats) {
    const { compilation } = stats;
    const files = [];
    // adds all chunks to the list of emitted files such as lazy loaded modules
    for (const chunk of compilation.chunks) {
        for (const file of chunk.files) {
            files.push({
                // The id is guaranteed to exist at this point in the compilation process
                // tslint:disable-next-line: no-non-null-assertion
                id: chunk.id.toString(),
                name: chunk.name,
                file,
                extension: (0, path_1.extname)(file),
                initial: chunk.isOnlyInitial(),
            });
        }
    }
    // other all files
    for (const file of Object.keys(compilation.assets)) {
        files.push({
            file,
            extension: (0, path_1.extname)(file),
            initial: false,
            asset: true,
        });
    }
    // dedupe
    return files.filter(({ file, name }, index) => files.findIndex((f) => f.file === file && (!name || name === f.name)) ===
        index);
}
exports.getEmittedFiles = getEmittedFiles;
